// tsuml2/dist/demo/katana.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlackMagicKatana = exports.MagicKatana = exports.Katana = exports.BaseWeapon = void 0;
const equipment_1 = __importDefault(require("./equipment"));
class BaseWeapon extends equipment_1.default {
    constructor() {
        super(...arguments);
        this.damage = 25;
        this.durability = 100;
        this.magicDurability = {
            fire: 100,
            water: 100,
        };
        this.attributes = [{ id: "explosive", value: 10 }, { id: "bouncing", value: 20 }]; // Test member Alias (import as)
    }
    refresh() {
        // DO Something 
    }
}
exports.BaseWeapon = BaseWeapon;
class Katana extends BaseWeapon {
    constructor() {
        super(...arguments);
        this.name = "Katana";
    }
    tryHit(fromDistance) {
        return fromDistance <= 2;
    }
    doIt() { }
}
exports.Katana = Katana;
class MagicKatana extends Katana {
    constructor(magic) {
        super();
        this.magic = magic;
    }
    tryMagicHit() {
        throw new Error("Method not implemented.");
    }
}
exports.MagicKatana = MagicKatana;
class BlackMagicKatana extends MagicKatana {
    tryBlackMagicHit() {
        throw new Error("Method not implemented.");
    }
}
exports.BlackMagicKatana = BlackMagicKatana;
