// tsuml2/dist/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TsUML2Settings = exports.postProcessSvg = exports.renderNomnomlSVG = exports.parseProject = exports.getNomnomlDSL = exports.getMermaidDSL = exports.createDiagram = void 0;
var core_1 = require("../core");
Object.defineProperty(exports, "createDiagram", { enumerable: true, get: function () { return core_1.createDiagram; } });
Object.defineProperty(exports, "getMermaidDSL", { enumerable: true, get: function () { return core_1.getMermaidDSL; } });
Object.defineProperty(exports, "getNomnomlDSL", { enumerable: true, get: function () { return core_1.getNomnomlDSL; } });
Object.defineProperty(exports, "parseProject", { enumerable: true, get: function () { return core_1.parseProject; } });
var io_1 = require("../core/io");
Object.defineProperty(exports, "renderNomnomlSVG", { enumerable: true, get: function () { return io_1.renderNomnomlSVG; } });
var emitter_1 = require("../core/emitter");
Object.defineProperty(exports, "postProcessSvg", { enumerable: true, get: function () { return emitter_1.postProcessSvg; } });
var tsuml2_settings_1 = require("../core/tsuml2-settings");
Object.defineProperty(exports, "TsUML2Settings", { enumerable: true, get: function () { return tsuml2_settings_1.TsUML2Settings; } });
