// tsuml2/dist/core/tsuml2-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TsUML2Settings = void 0;
class TsUML2Settings {
    constructor() {
        /**
         * the glob pattern defining a scope for typescript files to include (i.e.: a subfolder of the project)
         * required
         */
        this.glob = "";
        /**
         * the svg output file (with relative or absolute path)
         */
        this.outFile = "out.svg";
        /**
         * show property types
         */
        this.propertyTypes = true;
        /**
         * show private, protected, public, static modifiers
         */
        this.modifiers = true;
        /**
         * add type links
         */
        this.typeLinks = true;
        /**
         * nomnoml layouting and styling options
         */
        this.nomnoml = [];
        /**
         * output file contain the DSL (nomnoml)
         */
        this.outDsl = "";
        /**
         * output file containing mermaid DSL
         */
        this.outMermaidDsl = "";
        /**
         * mermaid layouting and styling options
         */
        this.mermaid = [];
        /**
         * show associations between classes, interfaces, types and their member types
         */
        this.memberAssociations = false;
        /**
         * show only exported types, classes, interfaces, enums
         */
        this.exportedTypesOnly = false;
    }
}
exports.TsUML2Settings = TsUML2Settings;
/*
    function isUserSetOption(option: string,yargs: Argv<{}>) {
        function searchForOption(option:string) {
          return (process.argv.indexOf(option) > -1);
        }
    
        if (searchForOption(`-${option}`) || searchForOption(`--${option}`))
          return true;
    
        // Handle aliases for same option
        for (let aliasIndex in yargs.choices().parsed.aliases[option]) {
          let alias = yargs.choices().parsed.aliases[option][aliasIndex];
          if (searchForOption(`-${alias}`) || searchForOption(`--${alias}`))
            return true;
        }
    
        return false;
    }
*/
